apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+                    {
+                        // Throw it.
+                        throw writeTask.Exception.FlattenIfPossible();
+                    }
+                }
+                else
+                {
+                    // Everything succeeded. Update _retryableIds dictionary.
+                    UpdateRetryableIds(ids, default);
+                }
+
+                // Perform push so that other messages can be processed.
+                StartPush();
+
+                // This method schedules the retry of ackIds/nackIds/extendIds with temporary errors and
+                // handles the permanent errors. Permanent errors are propagated to the caller for Acknowledge RPC,
+                // while permanent errors are only logged for ModifyAckDeadline RPC as ModifyAckDeadline call is not user initiated.
+                void RetryAcksAndExtends(bool? ack, AckError ackError, IEnumerable<TimedId> extendIds)
+                {
+                    // ack = true implies acks. ack = false implies nacks. ack = null implies extends, in which case extendIds should be non-empty.
+                    // Handle extends.
+                    bool hasExtends = extendIds.Any();
+                    if (hasExtends)
+                    {
+                        var extendIdsToRetry = extendIds.Where(j => ackError.TemporaryFailureIds.Contains(j.Id));
+                        if (extendIdsToRetry.Any())
+                        {
+                            // Retry for 10 seconds or 3 attempts only.
+                            RetryTemporaryFailures(Enumerable.Empty<string>(), null, extendIdsToRetry, extendsToRetry => _extendQueue.Locked(() => _extendQueue.Requeue(extendsToRetry)), 10);
+                        }
+                    }
+                    else
+                    {
+                        // We have acks or nacks.
+                        var idsToRetry = ackError.TemporaryFailureIds;
+                        var queue = ack.HasValue && ack.Value ? _ackQueue : _nackQueue;
+                        if (idsToRetry.Any())
+                        {
+                            // Retry for 600 seconds.
+                            RetryTemporaryFailures(idsToRetry, acksToRetry => queue.Locked(() => queue.Requeue(acksToRetry)), Enumerable.Empty<TimedId>(), null, 600);
+                        }
+                    }
+
+                    if (ackError.TryGetAcknowledgementException(out var exception))
+                    {
+                        if (hasExtendIds)
+                        {
+                            // Don't bubble permanent failures for lease extends as they are not user initiated.
+                            // Just log them.
+                            // TODO: Log the permanent failures.
+                            return;
+                        }
+
+                        // For acks/nacks, bubble up the exception.
+                        throw exception.FlattenIfPossible();
```

Comment:
If we're going to call `FlattenIfPossible`, maybe it would be simpler just to return a `List<AcknowledgementException>` from `TryGetPermanentException` (which would then be `GetPermanentExceptions`, and could return an empty list if there aren't any).